### ISSUE TITLE
CA1508 false positive in declaration pattern

### ISSUE THREAD

**Issue by ineternet** [dotnet/roslyn-analyzers#7164]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)
**Version**: [SDK 8.0.101](https://dotnet.microsoft.com/en-us/download/dotnet/8.0)

### Describe the bug

The following marks the final line as dead code (CA1508):
```cs
IComparable? firstNumber = null;
foreach (int item in new int[] { 1, 2, 3 })
    if (item is IComparable { } i)
        firstNumber ??= i; //'firstNumber is null' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
```
Different way to write it, same error:
```cs
IComparable? firstNumber = null;
foreach (int item in new int[] { 1, 2, 3 })
    if (item is IComparable { } i)
        if (firstNumber is null) //CA1508 happens here
            firstNumber = i;
```
The following alternatives, while functionally the same, do NOT lead to a false positive (they are not reported):
```cs
    if (item is int { } i)
    if (item is int i)
    if (item is IComparable i)
    if (item is { } i)
```
It only happens when the local variable is used. For example, the following does NOT produce the warning either:
```cs
IComparable? firstNumber = null;
foreach (int item in new int[] { 1, 2, 3 })
    if (item is IComparable { } i)
        if (firstNumber is null)
            firstNumber = item;
```
